eth-custodian/test/EthCustodianProxy.js

```diff
@@ -0,0 +1,266 @@
+                .to.be.revertedWith('Pausable: paused');
+        });
+
+        it('Should pause withdraw pre-migration', async () => {
+            await ethCustodianProxy.migrateToNewProofProducer(newProofProducerData, migrationBlock);
+            await ethCustodianProxy.pauseProxy(PAUSED_WITHDRAW_PRE_MIGRATION);
+            const proof = require('./proof_template_from_testnet.json');
+
+            await expect(ethCustodianProxy.withdraw(borshifyOutcomeProof(proof), 1099))
+                .to.be.revertedWith('Pausable: paused');
+        });
+
+        it('Should pause all', async () => {
+            await ethCustodianProxy.migrateToNewProofProducer(newProofProducerData, migrationBlock);
+            await ethCustodianProxy.pauseAll();
+
+            await expect(ethCustodianProxy.depositToNear('recipient.near', { value: 50000 }))
+                .to.be.revertedWith('Pausable: paused');
+
+            await expect(ethCustodianProxy.depositToEVM(ethRecipientOnNear.address, { value: 50000 }))
+                .to.be.revertedWith('Pausable: paused');
+
+            const proof = require('./proof_template_from_testnet.json');
+
+            await expect(ethCustodianProxy.withdraw(borshifyOutcomeProof(proof), 1099))
+                .to.be.revertedWith('Pausable: paused');
+
+            await expect(ethCustodianProxy.withdraw(borshifyOutcomeProof(proof), migrationBlock + 1))
+                .to.be.revertedWith('Pausable: paused');
+        });
+    });
+
+    describe('Migrate', () => {
+        it('Should change proof producer for EthCustodian', async () => {
+            const oldProofProducer = await ethCustodian.nearProofProducerAccount_();
+
+            await ethCustodianProxy.migrateToNewProofProducer(newProofProducerData, migrationBlock);
+
+            expect(await ethCustodianProxy.migrationBlockHeight())
+                .to.equal(migrationBlock);
+
+            expect(await ethCustodianProxy.preMigrationProducerAccount())
+                .to.equal(oldProofProducer);
+
+            expect(await ethCustodian.nearProofProducerAccount_())
+                .to.equal('0x' + newProofProducerData.toString('hex'));
+        });
+
+        it('Should fail when invoked for the second time', async () => {
+            await ethCustodianProxy.migrateToNewProofProducer(newProofProducerData, migrationBlock);
+            
+            await expect(ethCustodianProxy.migrateToNewProofProducer(newProofProducerData, migrationBlock))
+                .to.be.revertedWith('AlreadyMigrated');
+        });
+
+        it('Should fail when block producer id is too long', async () => {            
+            await expect(
+                ethCustodianProxy.migrateToNewProofProducer(Buffer.from('new-loooooooong-producer.testnet'), migrationBlock)
+            )
+                .to.be.revertedWith('ProducerAccountIdTooLong');
+        });
+    });
+
+    describe('Withdraw', () => {
```

Comment:
Extra tests:
- withdraw with blockHeight = migrationBlockNumber
- withdraw by old producer account after migration block number
- withdraw by new producer account before migration block number

Reply:
and double withdraw in one test (1) after migration block number, (2) before migration block number

Reply:
Agree with blockHeight = migrationBlockNumber

Not sure about the others. We already know from existing tests that correct producer is used depending on the block number. 2,3,4 will be testing the same thing